src/include/index/ivf_pq_index.h

```diff
@@ -813,6 +839,37 @@ class ivf_pq_index {
     return pq_vectors;
   }
 
+  template <
+      feature_vector_array U,
+      class Matrix,
+      class Distance = uncached_sub_sum_of_squares_distance>
+  auto pq_query_centroid_distances(
```

Comment:
Might be nice to add a comment like:

> For each vector, we will iterate through each of it's subspaces, and for each subspace we will compute the distance from the vector's subspace to each of the (256) clusters we've created for that subspace. We then store that distance, effectively caching the distance computation. Because we've computed this, later on, when we have a another vector which has been PQ-encoded and we want to compute the distance from our vector to this PQ-encoded vector, all we need to do is, for each subspace:
> 1. Get the cluster which the PQ-encoded vector's subspace is closest to.
> 2. Look in our pq_query_centroid_distances matrix for that cluster for this subspace. The value is a distance which we add to our total distance.
> 3. Continue this for all subspaces.

Though please feel free to edit / move somewhere else.

Reply:
Added more documentation